### PR TITLE
Update segmentation metrics

### PR DESCRIPTION
**Pull request by Dokholyan**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/metrics/region_base_metrics.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+from catalyst.utils.torch import get_activation_fn
+
+
+def _get_segmentation_stats(
+    outputs: torch.Tensor,
+    targets: torch.Tensor,
+    class_dim: Optional[int] = None,
+    threshold: float = None,
+    activation: str = "Sigmoid",
```
Comment (Scitator) [inline catalyst/metrics/region_base_metrics.py]: let's remove `activation` from all metrics
it's better to use extra wrappers for activations and work with correct `scores` and `targets` for metrics computation

File: catalyst/metrics/functional.py [hunk 2]
```diff
@@ -493,6 +493,37 @@ def topk_metric_with_dict_output(*args, **kwargs):
     return topk_metric_with_dict_output
 
 
+def get_aggregated_metric(per_class_metric: torch.Tensor,
```
Comment (Scitator) [inline catalyst/metrics/functional.py]: already have one :)
https://github.com/catalyst-team/catalyst/blob/master/catalyst/metrics/functional.py#L422

File: catalyst/metrics/region_base_metrics.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    dims.pop(class_dim)
+
+    sum_per_class = partial(torch.sum, dim=dims)
+
+    tp = sum_per_class(outputs * targets)
+    class_union = sum_per_class(outputs) + sum_per_class(targets)
+    class_union -= tp
+    fp = sum_per_class(outputs * (1 - targets))
+    fn = sum_per_class(targets * (1 - outputs))
+    return tp, fp, fn
```
Comment (Scitator) [inline catalyst/metrics/region_base_metrics.py:75]: and let's make the same API with `get_multiclass_statistics`

File: catalyst/metrics/region_base_metrics.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
+from typing import Optional, Tuple
+from functools import partial
+
+import torch
+
+
+def _get_segmentation_stats(
```
Comment (Scitator) [inline catalyst/metrics/region_base_metrics.py]: let's move it to `catalyst/metrics/funcitonal.py`

Comment (Scitator) [reply catalyst/metrics/region_base_metrics.py]: btw, could we use full naming?

Comment (Scitator) [reply catalyst/metrics/region_base_metrics.py]: and could you please add some tests for statistics computation?

**Review by Scitator — DISMISSED** [review]

File: catalyst/metrics/region_base_metrics.py (hunk 3 above)
Comment (Dokholyan) [reply catalyst/metrics/region_base_metrics.py:75]: Why should  we return tn and support, they are useless?

File: catalyst/contrib/nn/criterion/trevsky.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# flake8: noqa
+# @TODO: code formatting issue for 20.07 release
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/trevsky.py]: could you please remove this one?

File: catalyst/metrics/region_base_metrics.py (hunk 3 above)
Comment (Scitator) [reply catalyst/metrics/region_base_metrics.py:75]: why so? I think they could be useful for some other metrics and it's better to calculate them
or you think, that's too complicated right now?

**Comment by Dokholyan** [comment]
About statistics. I think that it is useless, I don't know any region-based segmentation metrics that use other statistics. So, I think,  your suggestion will make the function more complicated and that all

File: catalyst/callbacks/metrics/dice.py [hunk 6]
```diff
@@ -8,11 +9,12 @@
     calculate_tp_fp_fn,
 )
 from catalyst.core.callback import Callback, CallbackOrder
-from catalyst.metrics.dice import calculate_dice, dice
+from catalyst.metrics.dice import calculate_dice
 from catalyst.metrics.functional import (
     wrap_class_metric2dict,
     wrap_metric_fn_with_activation,
 )
+from catalyst.metrics.region_base_metrics import dice
```
Comment (Scitator) [inline catalyst/callbacks/metrics/dice.py:17]: btw, could you please move dice/iou/etc-statistic-based-metrics to `catalyst.metrics.classification`
quite simple logic – segmentation is just a per-pixel classification :)

File: catalyst/callbacks/metrics/iou.py [hunk 7]
```diff
@@ -28,23 +31,32 @@ def __init__(
...
+            class_dim: indicates class dimention (K) for
+                ``outputs`` and ``targets`` tensors (default = 1)
+            threshold: threshold for outputs binarization
             **kwargs: key-value params to pass to the metric
 
         .. note::
+<<<<<<< HEAD
+            For `**kwargs` info, please follow
+            `catalyst.metrics.region_base_metrics.iou` docs
+=======
```
Comment (Scitator) [inline catalyst/callbacks/metrics/iou.py]: fix this one please

File: catalyst/callbacks/metrics/iou.py [hunk 8]
```diff
@@ -61,7 +73,7 @@ def __init__(
...
 
 
-__all__ = [
-    "IouCallback",
-    "JaccardCallback",
-]
+# __all__ = [
+#     "IouCallback",
+#     "JaccardCallback",
+# ]
```
Comment (Scitator) [inline catalyst/callbacks/metrics/iou.py]: return this

File: catalyst/contrib/nn/criterion/dice.py [hunk 9]
```diff
@@ -1,44 +1,65 @@
 # flake8: noqa
 # @TODO: code formatting issue for 20.07 release
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/dice.py]: let's remove it

File: catalyst/contrib/nn/criterion/dice.py [hunk 10]
```diff
@@ -1,44 +1,65 @@
...
-        return 1 - dice
+    def forward(
+        self, outputs: torch.Tensor, targets: torch.Tensor
+    ) -> torch.Tensor:
+        """Calculates loss between ``logits`` and ``target`` tensors."""
+        dice_score = self.loss_fn(outputs, targets)
+        return 1 - dice_score
 
 
 class BCEDiceLoss(nn.Module):
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/dice.py]: would you like to remove this legacy loss?

File: catalyst/contrib/nn/criterion/iou.py [hunk 11]
```diff
@@ -1,44 +1,64 @@
 # flake8: noqa
 # @TODO: code formatting issue for 20.07 release
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/iou.py]: let's remove this one too

File: catalyst/contrib/nn/criterion/iou.py [hunk 12]
```diff
@@ -1,44 +1,64 @@
...
-        return 1 - iou
+    def forward(
+        self, outputs: torch.Tensor, targets: torch.Tensor
+    ) -> torch.Tensor:
+        """Calculates loss between ``logits`` and ``target`` tensors."""
+        iou_score = self.loss_fn(outputs, targets)
+        return 1 - iou_score
 
 
 class BCEIoULoss(nn.Module):
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/iou.py]: remove?

File: catalyst/contrib/nn/criterion/trevsky.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+    def forward(
+        self, outputs: torch.Tensor, targets: torch.Tensor
+    ) -> torch.Tensor:
+        """Calculates loss between ``logits`` and ``target`` tensors."""
+        trevsky_score = self.loss_fn(outputs, targets)
+        return 1 - trevsky_score
+
+
+__all__ = ["TrevskyLoss"]
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/trevsky.py]: btw, have you add it to `catalyst/contrib/nn/critetion/init.py`?

File: catalyst/metrics/dice.py [hunk 14]
```diff
@@ -113,4 +113,4 @@ def calculate_dice(
     return dice_metric
 
 
-__all__ = ["dice", "calculate_dice"]
```
Comment (Scitator) [inline catalyst/metrics/dice.py:116]: return the dice please, or remove it completely

File: catalyst/metrics/iou.py [hunk 15]
```diff
@@ -76,6 +76,6 @@ def iou(
     return iou_score
 
 
-jaccard = iou
-
-__all__ = ["iou", "jaccard"]
+# jaccard = iou
+#
+# __all__ = ["iou", "jaccard"]
```
Comment (Scitator) [inline catalyst/metrics/iou.py]: the same here - return, or remove the iou

File: catalyst/callbacks/metrics/dice.py (hunk 6 above)
Comment (Scitator) [reply catalyst/callbacks/metrics/dice.py:17]: but maybe we should move it to the next round of metrics updates

File: catalyst/metrics/region_base_metrics.py [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+        metric = torch.mean(metrics_per_class)
+
+    if mode == "weighted":
+        assert len(weights) == len(segmentation_stats[0])
+        metrics = torch.tensor(weights) * metrics_per_class
+        metric = torch.sum(metrics)
+
+    if mode == "separately":
+        metric = metrics_per_class
+
```
Comment (Scitator) [inline catalyst/metrics/region_base_metrics.py]: let's use if-else

File: catalyst/metrics/region_base_metrics.py [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+             mode='weighted', metric are calculated separately and than summed
+             over all classes with weights. If mode='separately', metric are
+             calculated separately for all classes
+        weights: class weights(for mode="weighted")
+
+    Returns:
+        Metric
+
+    """
+    assert mode in ["micro", "macro", "weighted", "separately"]
```
Comment (Scitator) [inline catalyst/metrics/region_base_metrics.py]: could we rename "separately" to "per-class" or something more class/label based?

Comment (Scitator) [reply catalyst/metrics/region_base_metrics.py]: classwise? labelwise?

File: catalyst/callbacks/metrics/dice.py (hunk 6 above)
Comment (Scitator) [reply catalyst/callbacks/metrics/dice.py:17]: update: yup, let's make it on another round

File: catalyst/metrics/region_base_metrics.py (hunk 17 above)
Comment (Scitator) [reply catalyst/metrics/region_base_metrics.py]: and here is the problem with micro-averaging:
```
suppose we have the following statistics/whatever
1, 2, 3, 4, 5

global mean
(1 + 2 + 3 + 4 + 5) / 5 -> micro-mean=3

per-batch mean:
(1 + 2 + 3) / 3 -> mean=2
(4 + 5) / 2 -> mean=4.5
(2 + 4.5) / 2 -> mean=3.25
```

Comment (Scitator) [reply catalyst/metrics/region_base_metrics.py]: the problem here is that because of the last batch (it could have a different size than the usual one), we could get biased estimate

File: catalyst/metrics/region_base_metrics.py [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+    class_union -= tp
+    fp = sum_per_class(outputs * (1 - targets))
+    fn = sum_per_class(targets * (1 - outputs))
+    return tp, fp, fn
+
+
+def _get_region_based_metrics(
+    outputs: torch.Tensor,
+    targets: torch.Tensor,
+    metric_function: Callable,
```
Comment (Scitator) [inline catalyst/metrics/region_base_metrics.py]: that's a bit weird, but could we use `metric_fn`? somehow such naming became a part of our codestyle

**Review by Scitator — DISMISSED** [review]

File: catalyst/callbacks/metrics/iou.py (hunk 8 above)
Comment (Dokholyan) [reply catalyst/callbacks/metrics/iou.py]: This is an old functions, it is better to remove them

File: catalyst/callbacks/metrics/dice.py [hunk 19]
```diff
@@ -66,100 +65,4 @@ def __init__(
         )
 
 
-class MultiClassDiceMetricCallback(Callback):
```
Comment (Scitator) [inline catalyst/callbacks/metrics/dice.py]: let's keep `MultiClassDiceMetricCallback`, your changes don't affect it, isn't it?

File: catalyst/contrib/nn/criterion/dice.py [hunk 20]
```diff
@@ -1,87 +1,64 @@
 # flake8: noqa
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/dice.py]:
```suggestion
```

File: catalyst/contrib/nn/criterion/iou.py [hunk 21]
```diff
@@ -1,78 +1,64 @@
 # flake8: noqa
 # @TODO: code formatting issue for 20.07 release
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/iou.py]:
```suggestion
```

File: catalyst/metrics/dice.py [hunk 22]
```diff
@@ -1,116 +0,0 @@
...
-
-    return dice_score
-
-
-# @TODO: remove
-def calculate_dice(
-    true_positives: np.array,
-    false_positives: np.array,
-    false_negatives: np.array,
-) -> np.array:
```
Comment (Scitator) [inline catalyst/metrics/dice.py:86]: for compatibility reasons, let's keep this one :)

**Comment by Dokholyan** [comment]
I need a little help with codestyle. I see ./catalyst/metrics/tests/test_region_statictics.py:6:1: D103 Missing docstring in public function, but all other functions also don't have docstrings. What I should do?
